syntax = "proto3";
package fraud_detection;

service FraudDetection {
  rpc CacheOrder (OrderRequest) returns (OrderResponse) {}  // 新加的缓存函数
  rpc CheckUserFraud (OrderRequest) returns (OrderResponse) {}  // 新加：检查用户欺诈
  rpc CheckCardFraud (OrderRequest) returns (OrderResponse) {}  // 新加：检查卡欺诈
  rpc ClearOrder (ClearRequest) returns (ClearResponse) {}  // 新加的：清理数据
}

message OrderRequest {
  string user_name = 1;
  string contact = 2;
  string card_number = 3;
  string expirationDate = 4;
  string cvv = 5;
  repeated Item items = 6;
  string user_comment = 7;
  BillingAddress billing_address = 8;
  Device device = 9;
  Browser browser = 10;
  string deviceLanguage = 11;
  string screenResolution = 12;
  string referrer = 13;
  repeated int32 vector_clock = 14;
  string order_id = 15;
}

message Item {
  string name = 1;
  int32 quantity = 2;
}

message BillingAddress {
  string street = 1;
  string city = 2;
  string state = 3;
  string zip = 4;
  string country = 5;
}

message Device {
  string type = 1;
  string model = 2;
  string os = 3;
}

message Browser {
  string name = 1;
  string version = 2;
}

message OrderResponse {
  bool success = 1;
  string message = 2;
  repeated int32 vector_clock = 3;
}

message ClearRequest {
    string order_id = 1;
    repeated int32 final_vector_clock = 2;  // VCf
}

message ClearResponse {
    bool success = 1;
    string message = 2;
}